File: src/main/kotlin/br/com/zup/beagleui/framework/action/Navigate.kt
package br.com.zup.beagleui.framework.action

enum class NavigatorEventType {
    OPEN_DEEP_LINK,
    ADD_VIEW,
    OPEN_VIEW,
    FINISH_VIEW,
    POP_VIEW
}

data class NavigatorData(
        val path: String,
        val data: Map<String, String>? = null
)

data class Navigate(
    val type: NavigatorEventType? = null /* = NavigatorEventType.ADD_VIEW */,
    val value: NavigatorData? = null
) : Action

File: Sources/BeagleUI/Sources/Navigation/Navigate.swift
//
//  Copyright © 08/11/19 Zup IT. All rights reserved.
//

/// Action to represent a screen transition
public enum Navigate: Action {
    
    var newPath: NewPath? {
        switch self {
        case .addView(let data), .swapView(let data), .presentView(let data):
            return NewPath(path: data.path, shouldPrefetch: data.shouldPrefetch)

        case .finishView, .popView, .popToView, .openDeepLink:
            return nil
        }
    }
    
    case openDeepLink(DeepLinkNavigation)
    
    case swapView(NewPath)
    case addView(NewPath)
    case presentView(NewPath)

    case finishView
    case popView
    case popToView(Path)

    public typealias Path = String
    public typealias Data = [String: String]

    public struct NewPath {
        public let path: Path
        public let shouldPrefetch: Bool
        
        public init(path: Path, shouldPrefetch: Bool = false) {
            self.path = path
            self.shouldPrefetch = shouldPrefetch
        }
    }
    
    public struct DeepLinkNavigation {
        public let path: Path
        public let data: Data?

        public init(path: Path, data: Data? = nil) {
            self.path = path
            self.data = data
        }
    }
}
